### DISCUSSION TITLE
Implementing math function in the Input Deck

### DISCUSSION THREAD

**Discussion by ppandit95** [peridigm/peridigm#164] [Q&A] [answered]
Dear Peridigm Community,

I am trying to implement a spatially varying elastic constant in the simulation using a single block in order to simulate a simple example with FGM.
Any help in this regard will be helpful as changing the constant value to a function in the input deck creates an error.

Regards
Pushkar

**Comment by johntfoster** [comment, marked as answer]
Function parsing is implemented for boundary conditions, but not for
material models. This example
<https://github.com/peridigm/peridigm/blob/master/examples/fragmenting_cylinder/fragmenting_cylinder.yaml>
demonstrates the use for boundary conditions. It should be straightforward
to look at the boundary conditions parsing code and add a material model
that can have functionally dependent properties.

**Comment by jritta** [comment]
Hi Pushkar,

my first guess would be to implement in the used material model something similar as it is done for the BC, so some `PG_RuntimeCompiler::Function>` have a look here: https://github.com/peridigm/peridigm/blob/8e6cd5f357bd6fd589d276f2d995bd1767bce975/src/core/Peridigm_BoundaryCondition.cpp#L80) 

Best Jonas

**Reply by ppandit95** [reply]
Hi Jonas

As you suggested I looked into the RTCompiler function but in order to implement it into the input parser I am not able to find the relevant C++ file as well as it would be quite beneficial to know how to use IDE with peridigm...

Any pointers in this regard will be helpful.

Many Thanks 
Pushkar

**Reply by djlittl** [reply]
The actual source for RTCompiler is in Trilinos under packages/pamgen/rtcompiler/

But I don't think you need to look into the source.

Have a look in the Peridigm file Peridigm_HorizonManager.cpp and search for "rtcFunction".  You can basically copy-and-paste this code.  There is a setup step and an evaluation step.  Examples of both steps are there in the Peridigm_HorizonManager.cpp file.

**Reply by ppandit95** [reply]
Thanks, @djlittl  for directing me to the relevant source I will certainly look into it moreover as Jonas has suggested Peridigm_BoundaryConditions.cpp has the setup for rtc function as well .....

**Reply by ppandit95** [reply]
As in the case of variable horizon function,it is computed during exodus mesh input using the global node ids before starting the simulation but in my case I am looking to implement a function in Peridigm_Materials.cpp which will compute the bulk/shear/young's modulus at each point during the run and return a computed double value instead of computing it once in the constructor for which Datamanagager could be used but I am confused about the data variables *x and *y which are double pointers but one takes in a scalar data and another one takes a vector data so in this regard a reference to the underlying data structure will be help full .

**Reply by oldninja** [reply]
> I am looking to implement a function in Peridigm_Materials.cpp which will compute the bulk/shear/young's modulus at each point during the run

@ppandit95, I am just trying to understand, you would be changing the material properties during the run? Or these material properties will be graded in a particular direction once at the beginning of the run? I am a bit confused.

**Comment by djlittl** [comment]
Another option that comes to mind is the Data Loader, which would allow you to put your spatially-varying data on the exodus/genesis mesh (perhaps with a python script using exodus.py) and read the data in at run time.  You would need to alter your material model code to access the corresponding data field and use it in the material model force evaluation.

See this test:


peridigm/test/regression/Notched_Plate_Data_Loader

For an example of how to read data into Peridigm from the mesh.

No idea if this is a viable option for you or not, just food for thought.

Dave

From: "John T. Foster" ***@***.***>
Reply-To: peridigm/peridigm ***@***.***>
Date: Monday, October 4, 2021 at 10:02 AM
To: peridigm/peridigm ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [EXTERNAL] Re: [peridigm/peridigm] Implementing math function in the Input Deck (Discussion #164)

Function parsing is implemented for boundary conditions, but not for
material models. This example
<https://github.com/peridigm/peridigm/blob/master/examples/fragmenting_cylinder/fragmenting_cylinder.yaml>
demonstrates the use for boundary conditions. It should be straightforward
to look at the boundary conditions parsing code and add a material model
that can have functionally dependent properties.

On Mon, Oct 4, 2021 at 10:56 AM Pushkar Anirudha Pandit <
***@***.***> wrote:

> Dear Peridigm Community,
>
> I am trying to implement a spatially varying elastic constant in the
> simulation using a single block in order to simulate a simple example with
> FGM.
> Any help in this regard will be helpful as changing the constant value to
> a function in the input deck creates an error.
>
> Regards
> Pushkar
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/peridigm/peridigm/discussions/164>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAEQR3MFJDUG37ZBQRE26WTUFHFC5ANCNFSM5FJWQB2Q>
> .
>


—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub<https://github.com/peridigm/peridigm/discussions/164#discussioncomment-1425588>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ABKXABILVAOD4G7X67WPPALUFHFZLANCNFSM5FJWQB2Q>.
Triage notifications on the go with GitHub Mobile for iOS<https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android<https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.

**Reply by ppandit95** [reply]
Thanks @djlittl , @jritta and @johntfoster for directing me in this context .

**Comment by ppandit95** [comment]
@oldninja for my implementation the properties will be graded at the
beginning of the simulation so I guess it will be a better idea to use data
loader as @djlittle suggested.

**Comment by ppandit95** [comment]
As suggested by David I tried Data Loader capability but it seems to create an error with an existing material model which when implemented says "2 material constants are required". Moreover, I am not sure whether the loaded fluid concentration is used in the computation process while calculating force ??
